lib/ansible/modules/network/dcnm/dcnm_vrf.py

```diff
@@ -37,237 +39,1042 @@
-        if response.get('vrfName') == vrf:
-            return True
 
-    if isinstance(response, list):
-        if response and response[0]:
-            if response[0].get('ERROR') == 'Bad Request':
-                return False
+    def diff_for_attach_deploy(self, want_a, have_a):
+
+        attach_list = list()
+
+        if not want_a:
+            return attach_list
+
+        dep_vrf = False
+        for want in want_a:
+            found = False
+            if have_a:
+                for have in have_a:
+                    if want['serialNumber'] == have['serialNumber']:
+                        found = True
+
+                        ## Is this allowed use case? User wants to update vlan on an already attached and deployed VRF?
+                        ## DCNM is accepting updates on an already deployed VRF attachment,
+                        ## Ex: update vlan ID on an already deployed VLAN, just need to re-attach with new vlan and re-deploy
+                        ## No need to detach and undeploy beforehand.
+                        if bool(have['isAttached']) and bool(want['isAttached']):
+                            if have['vlan'] != want['vlan']:
+                                del want['isAttached']
+                                attach_list.append(want)
+                                continue
+
+                        if bool(have['isAttached']) is not bool(want['isAttached']):
+                            del want['isAttached']
+                            attach_list.append(want)
+                            continue
+
+                        if bool(have['deployment']) is not bool(want['deployment']):
+                            dep_vrf = True
+
+            if not found:
+                if bool(want['deployment']):
+                    del want['isAttached']
+                    attach_list.append(want)
+
+        # logit(json.dumps(attach_list, indent=4, sort_keys=True))
+        return attach_list, dep_vrf
+
+
+    def update_attach_params(self, attach, vrf_name, deploy):
+        serial = ""
+        for ip, ser in self.ip_sn.items():
+            if ip == attach['ip_address']:
+                serial = ser
+
+        if attach:
+            attach.update({'fabric': self.fabric})
+            attach.update({'vrfName': vrf_name})
+            attach.update({'vlan': attach.get('vlan_id')})
+            attach.update({'deployment': deploy})
+            attach.update({'isAttached': deploy})
+            attach.update({'serialNumber': serial})
+            attach.update({'extensionValues': ""})
+            attach.update({'instanceValues': ""})
+            attach.update({'freeformConfig': ""})
+            if 'deploy' in attach:
+                del attach['deploy']
+            del attach['vlan_id']
+            del attach['ip_address']
+        else:
+            attach = dict()
+
+        return attach
+
+
+    def diff_for_create(self, want, have):
+        create = dict()
```

Comment:
I would use a guard clause here again to avoid the outer `if/else` block.

```python
def diff_for_create(self, want, have):
    if not have:
        return want

    create = {}
    <> logic in else block here <>
    
    return create
```

Reply:
Done.